### PR TITLE
enable file locking for c++ fstream and ofstream. control fsync after…

### PR DESCRIPTION
**Pull request by JaeseungYeom**
1. enable file locking for c++ fstream and ofstream
3. control fsync after producer write via environment variable
4. change the order of lock release such that it comes just before closing a file locked.
5. change the function name dyad_fetch to dyad_fetch_metadata
6. added error check for dyad_fetch_metadata with shared_storage enabled
7. suppress warning about usleep

### CODE REVIEW

File: include/dyad/stream/dyad_stream_api.hpp [hunk 1]
```diff
@@ -70,11 +71,17 @@ void fsync_ofstream (std::basic_ofstream<_CharT, _Traits>& os)
         os.flush ();
         fsync (static_cast<my_filebuf&> (*os.rdbuf ()).handle ());
     }
+#else
+    if (os.is_open ()) {
+        os.flush ();
```
Comment (hariharan-devarajan) [inline include/dyad/stream/dyad_stream_api.hpp:76]: Why do we need to do `os.flush()`?

File: include/dyad/stream/dyad_stream_api.hpp [hunk 2]
```diff
@@ -88,8 +95,103 @@ void fsync_fstream (std::basic_fstream<_CharT, _Traits>& os)
...
+       public:
+        int handle ()
+        {
+            return this->_M_file.fd ();
+        }
+    };
+
+    if (os.is_open ()) {
+        int fd = static_cast<my_filebuf&> (*os.rdbuf ()).handle ();
+        core.file_lock (fd);
```
Comment (hariharan-devarajan) [inline include/dyad/stream/dyad_stream_api.hpp:122]: Is there a way to select which lock your applying? We should use shared locks for read and exclusive locks for writes,

Comment (hariharan-devarajan) [reply include/dyad/stream/dyad_stream_api.hpp:122]: ah i see u are only only applying for producer. I think, we should have different methods for shared or exclusive locks for clarity of API.

**Review by hariharan-devarajan — CHANGES_REQUESTED** [review]

File: include/dyad/stream/dyad_stream_api.hpp (hunk 1 above)
Comment (JaeseungYeom) [reply include/dyad/stream/dyad_stream_api.hpp:76]: To make sure that data is cleared out before starting the followup synchronization routine, which can happen before d'tor call.

File: include/dyad/stream/dyad_stream_api.hpp (hunk 2 above)
Comment (hariharan-devarajan) [reply include/dyad/stream/dyad_stream_api.hpp:122]: I know maybe u wanna say exclusive lock

File: include/dyad/stream/dyad_stream_api.hpp (hunk 1 above)
Comment (hariharan-devarajan) [reply include/dyad/stream/dyad_stream_api.hpp:76]: I would not do that always. As I said, we should avoid any kind of flush.

File: include/dyad/stream/dyad_stream_api.hpp (hunk 2 above)
Comment (JaeseungYeom) [reply include/dyad/stream/dyad_stream_api.hpp:122]: We can add a shared lock to ifstream. Also to the fstream with `m_core.is_dyad_consumer()`

Comment (hariharan-devarajan) [reply include/dyad/stream/dyad_stream_api.hpp:122]: We don't need that as the consume code already does shared lock.

File: include/dyad/stream/dyad_stream_api.hpp (hunk 1 above)
Comment (JaeseungYeom) [reply include/dyad/stream/dyad_stream_api.hpp:76]: For any fencing semantic, I think this is needed.

File: include/dyad/stream/dyad_stream_api.hpp (hunk 2 above)
Comment (JaeseungYeom) [reply include/dyad/stream/dyad_stream_api.hpp]: I agree. That is why I  did not add it.

File: include/dyad/stream/dyad_stream_api.hpp (hunk 1 above)
Comment (hariharan-devarajan) [reply include/dyad/stream/dyad_stream_api.hpp:76]: the session semantics is maintained by closing already, we don't need that for most file systems.

Comment (JaeseungYeom) [reply include/dyad/stream/dyad_stream_api.hpp:76]: This is decoupled from file close. This is a utility function like mem fence instruction. If you use it just before stream close, there will be nothing to flush at close.

Comment (hariharan-devarajan) [reply include/dyad/stream/dyad_stream_api.hpp:76]: So this wont be called automatically by close of stream? Users need to call it explicitly?

Comment (hariharan-devarajan) [reply include/dyad/stream/dyad_stream_api.hpp:76]: My understanding is that this would not be an expensive operation. Therefore its ok.

**Review by hariharan-devarajan — APPROVED** [review]
The changes look good now.

**Pull request merged**